vllm/entrypoints/openai/tool_parsers.py

```diff
@@ -0,0 +1,624 @@
+        handling tool calls and streaming. Has to be an instance method because it requires state - the current text/
+        tokens/diffs, but also the information about what has previously been parsed and extracted (see constructor)
+        """
+        raise NotImplementedError('AbstractToolParser.extract_tool_calls_streaming has not been implemented!')
+
+
+class MistralToolParser(ToolParser):
+    """
+    Tool call parser for Mistral 7B Instruct v0.3, intended for use with the examples/tool_chat_template_mistral.jinja
+    template. There are server IMPORTANT CAVEATS for this parser:
+        - The chat template is NOT official and does not work well if you try to get the model to call 2+ tools at once.
+            Stick to only one tool call per generation, as the chat template is not reliable with > 1 and the model
+            Will lose coherence.
+        - Mistral's tool call format, that this translates into an OpenAI format, uses SINGLE QUOTES which cannot be
+            parsed to JSON. To enable JSON parsing and serialization, we find-and-replace these with DOUBLE QUOTES. To
+            prevent tool call corruption / deserialization failure, ensure that your tool calls and in particular your
+            ARGUMENTS never contain single or double quotes except as JSON control characters.
+
+    Used when --enable-api-tools --enable-auto-tool-choice --tool-call-parser mistral are all set
+    """
+
+    # the bot_token is the token indicating tool call(s) follow. Tokens before this token will be parsed as content; and
+    # if not present, the entire response will be parsed as text content.
+    bot_token: str = '[TOOL_CALLS]'  # string literal
+    bot_token_id: int = 5  # token ID thereof from the models' tokenizer
+
+    @staticmethod
+    def extract_tool_calls(model_output: str) -> ExtractedToolCallInformation:
+        """
+        Extract the tool calls from a complete model response. Requires find-and-replacing single quotes with double
+        quotes for JSON parsing, make sure your tool call arguments don't ever include quotes!
+        """
+
+        # Get the tool call token from the tokenizer
+        if MistralToolParser.bot_token not in model_output:
+            return ExtractedToolCallInformation(
+                tools_called=False,
+                tool_calls=[],
+                content=model_output
+            )
+        else:
+            try:
+                # extract the token so we hopefully have a JSON string
+                raw_tool_call = (model_output
+                                 .replace(MistralToolParser.bot_token, '')  # remove BOT token
+                                 .replace("'", '"'))  # ... hack to parse broken mistral JSON
+                # load the JSON, and then use it to build the Function and Tool Call
+                function_call_arr = json.loads(raw_tool_call)
```

Comment:
Hello! When I test with a [quantized Mistral V3 model](https://huggingface.co/neuralmagic/Mistral-7B-Instruct-v0.3-GPTQ-4bit) and the example in examples/openai_chat_completion_client_with_tools.py, the `raw_tool_call` is

`"[TOOL_CALLS][{'name': 'get_current_weather', 'arguments': {'city': 'Dallas', 'state': 'TX', 'unit': 'fahrenheit'}}, {'name': 'get_current_weather', 'arguments': {'city': 'San Francisco', 'state': 'CA', 'unit': 'fahrenheit'}}]$$ In order to execute these functions, I need to provide the current weather API. Let me find the data for you.\n\nIn Dallas, Texas, the temperature is currently 89°F.\n\nIn San Francisco, California, the temperature is currently 64°F."`

So maybe due to loss of performance in quantization, the model generates some extra suffix after a valid json list. Is there anything that we could do here to find the JSON substring? In `Hermes2ProToolParser.extract_tool_calls` on line 414 there is regex search for finding the tool call definition. Could that regex based search/findall be more robust also in case of Mistral?

Reply:
This may be possible. Let me take a look. May be more difficult to do in the case of streaming.

Reply:
9c0e6d8a for non-streaming